File: android/app/src/main/java/ru/bgtu_voenmeh/zapara/data/api/TimetableGroupRequest.kt
package ru.bgtu_voenmeh.zapara.data.api

/** A stored local id and its known group name, resolved against one pinned catalog generation. */
data class TimetableGroupRequest(val id: String?, val name: String?) {
    fun resolve(groups: List<TimetableApiGroup>): TimetableApiGroup {
        val direct = id?.let { wanted -> groups.firstOrNull { it.id == wanted } }
        val known = name?.trim()?.takeIf { it.isNotEmpty() }
        if (direct != null && (known == null || direct.name.equals(known, ignoreCase = true))) return direct
        val wanted = known ?: id
        val matches = groups.filter { it.name.equals(wanted, ignoreCase = true) }.take(2)
        if (matches.size != 1) throw TimetableApiException(TimetableApiFailure.UnknownRequiredGroup)
        return matches.single()
    }
}

File: src/Vograph.Core/Models/TimetableGroupRequest.cs
namespace Vograph.Core.Models;

/// <summary>A saved local identifier plus its known university group name, resolved within one catalog generation.</summary>
public sealed record TimetableGroupRequest(string? Id, string? Name)
{
    public TimetableApiGroup Resolve(IEnumerable<TimetableApiGroup> groups)
    {
        var catalog = groups.ToArray();
        var direct = Id is null ? null : catalog.FirstOrDefault(g => g.Id == Id);
        var known = string.IsNullOrWhiteSpace(Name) ? null : Name.Trim();
        if (direct is not null && (known is null || direct.Name.Equals(known, StringComparison.OrdinalIgnoreCase))) return direct;
        var wanted = known ?? Id;
        var matches = catalog.Where(g => g.Name.Equals(wanted, StringComparison.OrdinalIgnoreCase)).Take(2).ToArray();
        if (matches.Length != 1) throw new TimetableApiException(TimetableApiFailure.UnknownRequiredGroup);
        return matches[0];
    }
}
